### PR TITLE
cli: add create_artifact

### PR DESCRIPTION
**Pull request by itayadler**
create_artifact takes a path and creates a gzipped
tarball out of it. It uses the ignore crate to
filter out .gitignore and .ignore blacklist rules

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -8,6 +8,11 @@ edition = "2018"
 
 [dependencies]
 clap = "2.33.0"
+tar = "0.4"
+walkdir = "2"
+ignore = "0.4.7"
+tempfile = "3.0.8"
```
Comment (itayadler) [inline cli/Cargo.toml]: The nice thing I learnt about this is the file is really removed once the program terminates

File: cli/Cargo.toml [hunk 2]
```diff
@@ -8,6 +8,11 @@ edition = "2018"
 
 [dependencies]
 clap = "2.33.0"
+tar = "0.4"
+walkdir = "2"
+ignore = "0.4.7"
```
Comment (itayadler) [inline cli/Cargo.toml]: super cool crate that supports the .gitignore man page spec and also adds supports for a .ignore file

File: cli/src/main.rs [hunk 3]
```diff
@@ -9,20 +10,25 @@ fn main() {
...
         .author(crate_authors!())
         .setting(AppSettings::ArgRequiredElseHelp)
-        .subcommand(login_sub_command());
+        .subcommand(login_sub_command())
+        .subcommand(package_sub_command());
 
     let matches = app.get_matches();
     match matches.subcommand_name() {
         Some("login") => println!("loginnnn",), // login
+        Some("package") => package::create_package("./".to_string(), "/tmp/package.tar.gz".to_string()), // login
```
Comment (itayadler) [inline cli/src/main.rs]: was used just to run this, eventually the publish command will use create_package

File: cli/src/package.rs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+fn save_buffer_to_file(buffer: Vec<u8>, path: String) {
+  let mut output_file = File::create(path).unwrap();
+  output_file.write_all(&buffer).unwrap();
+}
+
+#[test]
+fn should_create_package() {
+  let output_path = "/tmp/test_package.tar.gz";
+  create_package("./tests/fixtures/test_package".to_string(), output_path.to_string());
```
Comment (itayadler) [inline cli/src/package.rs]: I do to_string() everywhere but I need to read about strings to see what the fuck is going on with it

File: cli/src/main.rs [hunk 5]
```diff
@@ -1,5 +1,6 @@
 #[macro_use]
 extern crate clap;
+mod package;
```
Comment (BlueHotDog) [inline cli/src/main.rs]: does the ordering here matters? can you put Mod before extern? does it affect anything?

File: cli/src/main.rs (hunk 3 above)
Comment (BlueHotDog) [reply cli/src/main.rs]: Just some terminology nit picking, is a package the tar? or is that the artifact? how these concept map to each other in your opinion?

File: cli/src/main.rs [hunk 6]
```diff
@@ -9,20 +10,25 @@ fn main() {
...
 fn login_sub_command<'a, 'b>() -> clap::App<'a, 'b> {
     return SubCommand::with_name("login").about("Login to buff").arg(
         Arg::with_name("username")
             .short("u")
             .help("Case sensative username"),
     );
 }
+
+fn package_sub_command<'a, 'b>() -> clap::App<'a, 'b> {
+    return SubCommand::with_name("package").about("package");
```
Comment (BlueHotDog) [inline cli/src/main.rs]: Lets add help and other niceties to this..

File: cli/src/package.rs [hunk 7]
```diff
@@ -0,0 +1,58 @@
+extern crate tar;
+extern crate ignore;
+use std::path::Path;
```
Comment (BlueHotDog) [inline cli/src/package.rs]: holy crap. so many uses.

File: cli/src/main.rs (hunk 3 above)
Comment (BlueHotDog) [inline cli/src/main.rs]: i would assume we want to directly stream it, might be premature optimization, but i dont think a file is really needed.

File: cli/src/package.rs (hunk 4 above)
Comment (BlueHotDog) [reply cli/src/package.rs]: yes, please do and tell me..

File: cli/tests/fixtures/test_package/buff.toml [hunk 8]
```diff
@@ -0,0 +1,6 @@
+[package]
```
Comment (BlueHotDog) [inline cli/tests/fixtures/test_package/buff.toml]: i think it'll be very cool to do a full package spec, even if we are not intend to support everything from the get-go.

**Comment by BlueHotDog** [comment]
I think the followup tasks for this are:
- Make CircleCI work for the CLI
- Make Code Coverage work for CLI
- Spec out the package.toml file

wdyt?

File: cli/src/main.rs (hunk 5 above)
Comment (itayadler) [reply cli/src/main.rs]: it doesn't, I think we should add linters and use a style that seems the leading one

File: cli/src/main.rs (hunk 3 above)
Comment (itayadler) [reply cli/src/main.rs]: > i would assume we want to directly stream it, might be premature optimization, but i dont think a file is really needed.

Definitely agree, as I mentioned I saved it to a file as it helped me during the development of this, as a file is easier to debug and verify the behavior is what we want. (From outside the program)
I think we should keep it for debugging purposes, and once we know whats the interaction between the artifact and publishing, we should refactor this appropriately

File: cli/src/package.rs (hunk 7 above)
Comment (itayadler) [reply cli/src/package.rs]: yeah a lot of glue to make this happen :)

File: cli/src/main.rs (hunk 3 above)
Comment (itayadler) [reply cli/src/main.rs]: changed to artifact

File: cli/src/package.rs (hunk 4 above)
Comment (itayadler) [reply cli/src/package.rs]: changed to &str
https://hermanradtke.com/2015/05/03/string-vs-str-in-rust-functions.html

File: cli/src/main.rs (hunk 6 above)
Comment (itayadler) [reply cli/src/main.rs]: removed it, as package won't be a real command IMHO, so just used in the test atm

**Comment by itayadler** [comment]
> I think the followup tasks for this are:
> 
> * Make CircleCI work for the CLI
> * Make Code Coverage work for CLI
> * Spec out the package.toml file
> 
> wdyt?

I'll open tasks for it

**Pull request merged**